AsyncDisplayKit/Details/ASRangeController.mm

```diff
@@ -159,42 +82,64 @@ - (void)updateVisibleNodeIndexPaths
-    NSSet *visibleRangeIndexPaths = [NSSet setWithArray:indexPaths];
-
-    NSMutableSet *removedIndexPaths = _workingRangeIsValid ? [_workingRangeIndexPaths mutableCopy] : [NSMutableSet set];
-    [removedIndexPaths minusSet:workingRangeIndexPaths];
-    [removedIndexPaths minusSet:visibleRangeIndexPaths];
-    if (removedIndexPaths.count) {
-      NSArray *removedNodes = [_delegate rangeController:self nodesAtIndexPaths:[removedIndexPaths allObjects]];
-      [removedNodes enumerateObjectsUsingBlock:^(ASCellNode *node, NSUInteger idx, BOOL *stop) {
-        [self removeNodeFromWorkingRange:node];
-      }];
-    }
-
-    NSMutableSet *addedIndexPaths = [workingRangeIndexPaths mutableCopy];
-    [addedIndexPaths minusSet:_workingRangeIndexPaths];
-    [addedIndexPaths minusSet:visibleRangeIndexPaths];
-    if (addedIndexPaths.count) {
-      NSArray *addedNodes = [_delegate rangeController:self nodesAtIndexPaths:[addedIndexPaths allObjects]];
-      [addedNodes enumerateObjectsUsingBlock:^(ASCellNode *node, NSUInteger idx, BOOL *stop) {
-        [self addNodeToWorkingRange:node];
-      }];
+  // the layout controller needs to know what the current visible indices are to calculate range offsets
+  [_layoutController setVisibleNodeIndexPaths:visibleNodePaths];
+
+  for (NSInteger i = 0; i < ASLayoutRangeTypeCount; i++) {
+    ASLayoutRangeType rangeType = (ASLayoutRangeType)i;
+    id rangeKey = @(rangeType);
+
+    // this delegate decide what happens when a node is added or removed from a range
+    id<ASRangeDelegate> rangeDelegate = _rangeTypeDelegates[rangeKey];
+
+    if ([_layoutController shouldUpdateForVisibleIndexPaths:visibleNodePaths viewportSize:viewportSize rangeType:rangeType]) {
+      NSSet *indexPaths = [_layoutController indexPathsForScrolling:_scrollDirection viewportSize:viewportSize rangeType:rangeType];
+
+      // Notify to remove indexpaths that are leftover that are not visible or included in the _layoutController calculated paths
+      NSMutableSet *removedIndexPaths = _rangeIsValid ? [[_rangeTypeIndexPaths objectForKey:rangeKey] mutableCopy] : [NSMutableSet set];
+      [removedIndexPaths minusSet:indexPaths];
+      [removedIndexPaths minusSet:visibleNodePathsSet];
+      if (removedIndexPaths.count) {
+        NSArray *removedNodes = [_delegate rangeController:self nodesAtIndexPaths:[removedIndexPaths allObjects]];
+        [removedNodes enumerateObjectsUsingBlock:^(ASCellNode *node, NSUInteger idx, BOOL *stop) {
+          [rangeDelegate node:node exitedRangeType:rangeType];
+        }];
+      }
+
+      // Notify to add indexpaths that are not currently in _rangeTypeIndexPaths
+      NSMutableSet *addedIndexPaths = [indexPaths mutableCopy];
+      [addedIndexPaths minusSet:[_rangeTypeIndexPaths objectForKey:rangeKey]];
+
+      // The preload range (for example) should include nodes that are visible
+      if ([self shouldRemoveVisibleNodesFromRangeType:rangeType]) {
```

Comment:
add a TODO to remove this special case once we decouple node rendering from Core Animation?